In [1]:
#import the necessities
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import statsmodels.api as sm
import warnings
from sqlalchemy import create_engine
warnings.filterwarnings('ignore')

1. Weather model
First, load the dataset from the weatherinszeged table from Thinkful's database.

In [2]:
#create and dispose of a connection for weather
postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
weather_db = 'weatherinszeged'

engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, weather_db))

weather_df = pd.read_sql_query('select * from weatherinszeged',con=engine)

engine.dispose()

As in the previous checkpoint, build a linear regression model where your target variable is the difference between the apparenttemperature and the temperature. As explanatory variables, use humidity and windspeed. Now, estimate your model using OLS. What are the R-squared and adjusted R-squared values? Do you think they are satisfactory? Why?

In [3]:
#create a target dataset
weather_y = (weather_df['apparenttemperature'] - weather_df['temperature'])
#create a feature dataset
weather_X = weather_df[['humidity', 'windspeed']]
#and add a constant
weather_X = sm.add_constant(weather_X)

In [4]:
#now run a OLS and get the summary
weather_results = sm.OLS(weather_y, weather_X).fit()
display(weather_results.summary())

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Thu, 14 Jan 2021   Prob (F-statistic):               0.00
Time:                        15:51:59   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.4381      0.021    115.948      0.000       2.397       2.479
humidity      -3.0292      0.024   -126.479      0.000      -3.076      -2.982
windspeed     -0.1193      0.001   -176.164      0.000      -0.121      -0.118
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.267
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                          -0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

The R-squared and Adj. R-squared values are 0.288. They are far too low to be satisfactory.

Next, include the interaction of humidity and windspeed to the model above and estimate the model using OLS. Now, what is the R-squared of this model? Does this model improve upon the previous one?

In [5]:
#add the interaction to the feature dataset
##while there is no need to recreate the target dataset
###it's practice for me to continue doing so
weather_i_y = (weather_df['apparenttemperature'] - weather_df['temperature'])
weather_i_X = weather_df[['humidity', 'windspeed']]
weather_i_X['humidity windspeed interaction'] = (weather_df['humidity'] * weather_df['windspeed'])
weather_i_X = sm.add_constant(weather_i_X)

In [6]:
#run a OLS with the new sets
weather_i_results = sm.OLS(weather_i_y, weather_i_X).fit()
display(weather_i_results.summary())

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Thu, 14 Jan 2021   Prob (F-statistic):               0.00
Time:                        16:01:30   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==================================================================================================
                                     coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
const                              0.0839      0.033      2.511      0.012       0.018       0.149
humidity                           0.1775      0.043      4.133      0.000       0.093       0.262
windspeed                          0.0905      0.002     36.797      0.000       0.086       0.095
humidity windspeed interaction    -0.2971      0.003    -88.470      0.000      -0.304      -0.291
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.265
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                          -0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

The R-squared and Adj. R-squared values are now 0.341. Better, but still too low to be satisfactory.

Add visibility as an additional explanatory variable to the first model and estimate it. Did R-squared increase? What about adjusted R-squared? Compare the differences put on the table by the interaction term and the visibility in terms of the improvement in the adjusted R-squared. Which one is more useful?

In [7]:
#create a target dataset with visibility included
weather_v_y = (weather_df['apparenttemperature'] - weather_df['temperature'])
weather_v_X = weather_df[['humidity', 'windspeed', 'visibility']]
weather_v_X = sm.add_constant(weather_v_X)

In [8]:
#run OLS once again
weather_v_results = sm.OLS(weather_v_y, weather_v_X).fit()
display(weather_v_results.summary())

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.304
Model:                            OLS   Adj. R-squared:                  0.303
Method:                 Least Squares   F-statistic:                 1.401e+04
Date:                Thu, 14 Jan 2021   Prob (F-statistic):               0.00
Time:                        16:06:48   Log-Likelihood:            -1.6938e+05
No. Observations:               96453   AIC:                         3.388e+05
Df Residuals:                   96449   BIC:                         3.388e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.5756      0.028     56.605      0.000       1.521       1.630
humidity      -2.6066      0.025   -102.784      0.000      -2.656      -2.557
windspeed     -0.1199      0.001   -179.014      0.000      -0.121      -0.119
visibility     0.0540      0.001     46.614      0.000       0.052       0.056
==============================================================================
Omnibus:                     3833.895   Durbin-Watson:                   0.282
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4584.022
Skew:                          -0.459   Prob(JB):                         0.00
Kurtosis:                       3.545   Cond. No.                         131.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

The R-squared value is 0.304 and Adj. R-squared value is 0.303. Better than model one but worse than model two.

Choose the best one from the three models above with respect to their AIC and BIC scores. Validate your choice by discussing your justification with your mentor.

The first model's AIC and BIC scores are 3.409e+05, the second model's AIC and BIC scores are 3.334e+05, and the third model's AIC and BIC scores are 3.388e+05. The second model has the lowest scores meaning it is the most useful model.

2. House prices model
Load the houseprices data from Thinkful's database.

In [9]:
#create and dispose of a connection for houseprices
postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
houses_db = 'houseprices'

engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, houses_db))

houses_df = pd.read_sql_query('select * from houseprices',con=engine)

engine.dispose()

Run your house prices model again and assess the goodness of fit of your model using an F-test, R-squared, adjusted R-squared, AIC, and BIC.

In [10]:
#create the target dataset and the feature dataset
##the first features used were lotarea, overallqual, overallcond,
## and centralair  but overallcond was dropped due to a low p-value
houses_y = houses_df['saleprice']
dummy_df = pd.get_dummies(houses_df['centralair'], drop_first=True)
houses_X = houses_df[['lotarea', 'overallqual']].merge(dummy_df,
                   left_index=True, right_index=True)
houses_X = sm.add_constant(houses_X)

In [11]:
#run the OLS
houses_results = sm.OLS(houses_y, houses_X).fit()
display(houses_results.summary())

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.660
Model:                            OLS   Adj. R-squared:                  0.659
Method:                 Least Squares   F-statistic:                     940.5
Date:                Thu, 14 Jan 2021   Prob (F-statistic):               0.00
Time:                        16:16:33   Log-Likelihood:                -17757.
No. Observations:                1460   AIC:                         3.552e+04
Df Residuals:                    1456   BIC:                         3.554e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -1.119e+05   6424.464    -17.411      0.000   -1.24e+05   -9.93e+04
lotarea         1.4442      0.122     11.798      0.000       1.204       1.684
overallqual  4.379e+04    916.959     47.753      0.000     4.2e+04    4.56e+04
Y            1.125e+04   5117.296      2.198      0.028    1208.844    2.13e+04
==============================================================================
Omnibus:                      591.663   Durbin-Watson:                   1.974
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             7628.923
Skew:                           1.522   Prob(JB):                         0.00
Kurtosis:                      13.777   Cond. No.                     8.62e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 8.62e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Do you think your model is satisfactory? If so, why?

It is far better than the weather models above but I would perfer to have a higher Adj. R-squared value. Prob (F-statistic) is great and the AIC and BIC are far lower than the weather models above.

In order to improve the goodness of fit of your model, try different model specifications by adding or removing some variables.
For each model you try, get the goodness of fit metrics and compare your models with each other. Which model is the best and why?

In [16]:
#we'll add a few features; specifically yearbuilt, yrsold, poolarea
houses_2_y = houses_df['saleprice']
dummy_2_df = pd.get_dummies(houses_df['centralair'], drop_first=True)
houses_2_X = houses_df[['lotarea', 'overallqual', 'yearbuilt', 'yrsold',
                   'poolarea']].merge(dummy_2_df, left_index=True, right_index=True)
houses_2_X = sm.add_constant(houses_2_X)

In [17]:
#third verse same as the first
houses_2_results = sm.OLS(houses_2_y, houses_2_X).fit()
display(houses_2_results.summary())

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.669
Model:                            OLS   Adj. R-squared:                  0.667
Method:                 Least Squares   F-statistic:                     488.9
Date:                Thu, 14 Jan 2021   Prob (F-statistic):               0.00
Time:                        16:38:26   Log-Likelihood:                -17737.
No. Observations:                1460   AIC:                         3.549e+04
Df Residuals:                    1453   BIC:                         3.553e+04
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const        -2.55e+05   1.82e+06     -0.140      0.889   -3.82e+06    3.31e+06
lotarea         1.4742      0.121     12.140      0.000       1.236       1.712
overallqual  4.026e+04   1070.108     37.619      0.000    3.82e+04    4.24e+04
yearbuilt     306.5340     50.696      6.047      0.000     207.089     405.979
yrsold       -214.8978    905.068     -0.237      0.812   -1990.278    1560.483
poolarea       62.1689     30.065      2.068      0.039       3.193     121.145
Y            2049.0005   5277.760      0.388      0.698   -8303.843    1.24e+04
==============================================================================
Omnibus:                      604.410   Durbin-Watson:                   1.984
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             8264.769
Skew:                           1.546   Prob(JB):                         0.00
Kurtosis:                      14.238   Cond. No.                     2.22e+07
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.22e+07. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

yrsold and centralair have very high p-values. We'll run one final one without those variables.

In [15]:
#create the datasets once more
houses_3_y = houses_df['saleprice']
houses_3_X = houses_df[['lotarea', 'overallqual', 'yearbuilt', 'poolarea']]
houses_3_X = sm.add_constant(houses_3_X)

In [18]:
#99 bottles of OLS on the wall, 99 bottles of OLS...
houses_3_results = sm.OLS(houses_3_y, houses_3_X).fit()
display(houses_3_results.summary())

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.669
Model:                            OLS   Adj. R-squared:                  0.668
Method:                 Least Squares   F-statistic:                     734.2
Date:                Thu, 14 Jan 2021   Prob (F-statistic):               0.00
Time:                        16:39:43   Log-Likelihood:                -17738.
No. Observations:                1460   AIC:                         3.549e+04
Df Residuals:                    1455   BIC:                         3.551e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const        -6.96e+05   9.21e+04     -7.555      0.000   -8.77e+05   -5.15e+05
lotarea         1.4763      0.121     12.174      0.000       1.238       1.714
overallqual  4.029e+04   1067.033     37.757      0.000    3.82e+04    4.24e+04
yearbuilt     312.2145     48.510      6.436      0.000     217.057     407.372
poolarea       62.6893     29.996      2.090      0.037       3.850     121.529
==============================================================================
Omnibus:                      602.355   Durbin-Watson:                   1.983
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             8236.174
Skew:                           1.539   Prob(JB):                         0.00
Kurtosis:                      14.221   Cond. No.                     1.12e+06
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.12e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

p-values are all < 0.05, Adj. R-squared value is slightly better, and AIC and BIC scores are either the same or slightly lower. This model is the best so far.